src/System.Collections.Immutable/tests/ImmutableArrayTest.cs

```diff
@@ -20,239 +23,326 @@ public class ImmutableArrayTest : SimpleElementImmutablesTestBase
 
-            var copy2 = ImmutableArray.CreateRange(array, i => i + 1);
-            Assert.Equal(new[] { 5, 6, 7, 8 }, copy2);
+            var array = new int[0];
+            Assert.IsType<ImmutableArray<int>>(ImmutableArray.Create(array));
 
-            Assert.Equal(new int[] { }, ImmutableArray.CreateRange(s_empty, i => i));
+            var immutable = ImmutableArray<int>.Empty;
+            Assert.IsType<ImmutableArray<ImmutableArray<int>>>(ImmutableArray.Create(immutable));
 
-            Assert.Throws<ArgumentNullException>("selector", () => ImmutableArray.CreateRange(array, (Func<int, int>)null));
+            var enumerable = Enumerable.Empty<int>();
+            Assert.IsType<ImmutableArray<IEnumerable<int>>>(ImmutableArray.Create(enumerable));
         }
 
         [Fact]
-        public void CreateRangeFromImmutableArrayWithSelectorAndArgument()
+        public void CreateEmpty()
         {
-            var array = ImmutableArray.Create(4, 5, 6, 7);
-
-            var copy1 = ImmutableArray.CreateRange(array, (i, j) => i + j, 0.5);
-            Assert.Equal(new[] { 4.5, 5.5, 6.5, 7.5 }, copy1);
-
-            var copy2 = ImmutableArray.CreateRange(array, (i, j) => i + j, 1);
-            Assert.Equal(new[] { 5, 6, 7, 8 }, copy2);
-
-            var copy3 = ImmutableArray.CreateRange(array, (int i, object j) => i, null);
-            Assert.Equal(new[] { 4, 5, 6, 7 }, copy3);
+            Assert.Equal(s_empty, ImmutableArray.Create<int>());
+            Assert.Equal(s_empty, ImmutableArray.Create(new int[0]));
+        }
 
-            Assert.Equal(new int[] { }, ImmutableArray.CreateRange(s_empty, (i, j) => i + j, 0));
+        [Theory]
+        [MemberData(nameof(Int32EnumerableData))]
+        public void CreateRange(IEnumerable<int> source)
+        {
+            Assert.Equal(source, ImmutableArray.CreateRange(source));
+        }
 
-            Assert.Throws<ArgumentNullException>("selector", () => ImmutableArray.CreateRange(array, (Func<int, int, int>)null, 0));
+        [Fact]
+        public void CreateRangeInvalid()
+        {
+            Assert.Throws<ArgumentNullException>("items", () => ImmutableArray.CreateRange((IEnumerable<int>)null));
         }
 
         [Fact]
-        public void CreateRangeSliceFromImmutableArrayWithSelector()
+        public void CreateRangeEmptyReturnsSingleton()
         {
-            var array = ImmutableArray.Create(4, 5, 6, 7);
+            var empty = ImmutableArray.CreateRange(new int[0]);
+            // This equality check returns true if the underlying arrays are the same instance.
+            Assert.True(s_empty == empty);
+        }
 
-            var copy1 = ImmutableArray.CreateRange(array, 0, 0, i => i + 0.5);
-            Assert.Equal(new double[] { }, copy1);
+        [Theory]
+        [MemberData(nameof(CreateRangeWithSelectorData))]
+        public void CreateRangeWithSelector<TResult>(IEnumerable<int> source, Func<int, TResult> selector, TResult dummy)
```

Comment:
Another xUnit issue that has been fixed, but not made it into corefx yet: the type of `TResult` cannot be inferred unless there is a `TResult` parameter that isn't wrapped. https://github.com/xunit/xunit/pull/965